**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №10

Набор данных: auto.csv

Атрибуты: 

### 1.Чтение файла (набора данных)

In [15]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
# чтение таблицы auto.csv
df = pd.read_csv('auto.csv')

Таким образом у нас теперь есть переменная df, которая хранит данные auto.csv в виде объекта DataFrame библиотеки pandas.

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [5]:
# применить метод head
df.head(20) # чтение первых 20-ти строк из df

Name  year  SellingPrice  kmdriven     fuel  \
0                     Maruti 800 AC  2007       60000.0   70000.0   Petrol   
1          Maruti Wagon R LXI Minor  2007      135000.0   50000.0   Petrol   
2              Hyundai Verna 1.6 SX  2012      600000.0  100000.0   Diesel   
3            Datsun RediGO T Option  2017      250000.0   46000.0   Petrol   
4             Honda Amaze VX i-DTEC  2014      450000.0  141000.0   Diesel   
5              Maruti Alto LX BSIII  2007      140000.0  125000.0   Petrol   
6         Hyundai Xcent 1.2 Kappa S  2016      550000.0   25000.0   Petrol   
7         Tata Indigo Grand Petroll  2014      240000.0   60000.0   Petrol   
8          Hyundai Creta 1.6 VTVT S  2015      850000.0   25000.0  Petroll   
9          Maruti Celerio Green VXI  2017      365000.0   78000.0      CNG   
10          Chevrolet Sail 1.2 Base  2015      260000.0   35000.0   Petrol   
11         Tata Indigo Grand Petrol  2014      250000.0  100000.0   Petrol   
12  Toyota Corolla Altis 1.8 VL CVT  2018     1650000.0   25000.0   Petrol   
13                    Maruti 800 AC  2007       60000.0   70000.0   Petrol   
14         Maruti Wagon R LXI Minor  2007      135000.0   50000.0   Petrol   
15             Hyundai Verna 1.6 SX  2012      600000.0  100000.0   Diesel   
16           Datsun RediGO T Option  2017      250000.0   46000.0   Petrol   
17            Honda Amaze VX i-DTEC  2014      450000.0  141000.0   Diesel   
18             Maruti Alto LX BSIII  2007      140000.0  125000.0   Petrol   
19        Hyundai Xcent 1.2 Kappa S  2016      550000.0   25000.0   Petrol   

   seller_Type transmission         owner  
0   Individual       Manual   First Owner  
1   Individual       Manual   First Owner  
2   Individual       Manual   First Owner  
3   Individual       Manual   First Owner  
4   Individual       Manual  Second Owner  
5   Individual       Manual   First Owner  
6   Individual       Manual   First Owner  
7   Individual       Manual  Second Owner  
8   Individual       Manual   First Owner  
9   Individual       Manual   First Owner  
10  Individual       Manual   First Owner  
11  Individual       Manual   First Owner  
12      Dealer    Automatic   First Owner  
13  Individual       Manual   First Owner  
14  Individual       Manual   First Owner  
15  Individual       Manual   First Owner  
16  Individual       Manual   First Owner  
17  Individual       Manual  Second Owner  
18  Individual       Manual   First Owner  
19  Individual       Manual   First Owner

2.2 Оценка данных с помощью метода info.

In [5]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4344 entries, 0 to 4343
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          4344 non-null   object 
 1   year          4344 non-null   int64  
 2   SellingPrice  4342 non-null   float64
 3   kmdriven      4343 non-null   float64
 4   fuel          4343 non-null   object 
 5   seller_Type   4344 non-null   object 
 6   transmission  4344 non-null   object 
 7   owner         4344 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 271.6+ KB


Метод info() в pandas используется для вывода краткой и полезной сводки о DataFrame.
Он позволяет быстро оценить структуру и качество данных, выявить пропущенные значения, проверить типы данных и понять объем занимаемой памяти. Метод особенно полезен на начальном этапе анализа данных для планирования дальнейших шагов обработки и исследования.

2.3 Оценка данных с помощью метода describe.

In [4]:
# оцените числовые столбцы с помощью describe

df.describe()

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


year  SellingPrice       kmdriven
count  4344.000000  4.342000e+03    4343.000000
mean   2013.092311  5.039987e+05   66193.063320
std       4.214543  5.784465e+05   46636.648764
min    1992.000000  2.000000e+04       1.000000
25%    2011.000000  2.100000e+05   35000.000000
50%    2014.000000  3.500000e+05   60000.000000
75%    2016.000000  6.000000e+05   90000.000000
max    2020.000000  8.900000e+06  806599.000000

Метод describe() используется для получения описательной статистики по данным DataFrame. Он вычисляет основные статистические показатели, которые помогают быстро понять распределение и характеристики данных.


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


Столбец year
- Диапазон лет автомобилей: от 1992 до 2020 года.
- Средний год выпуска: примерно 2013.
- Медиана (50%): 2014, то есть половина машин новее 2014 года, половина — старее.
- Стандартное отклонение (std): около 4.2, что говорит о довольно широком разбросе по годам производства.

Столбец SellingPrice
- Цены варьируются от 20 000 до 890 000.
- Средняя цена продажи около 504 000.
- Медианная продажная цена: 350 000.
- Первый квартиль (25%): 210 000, третий квартиль (75%): 650 000, что наглядно показывает распределение цен.
- Стандартное отклонение: 578 000, то есть цены сильно различаются между машинами.

Столбец kmdriven
- Средний пробег: 66 193 км.
- Медианный пробег: 60 000 км, типичный показатель для подержанных машин.
- Минимальный пробег: 1 км.
- Максимальный пробег: 806 599 км, что указывает на очень «прожившие» авто.
- Квартильные значения помогают увидеть распределение — 25% авто имеют пробег ниже 35 000 км, 75% — ниже 89 000 км.

Общие выводы
- Данные охватывают широкий диапазон лет, цен и пробегов.
- Большая вариативность по всем признакам.
- Медиана во всех случаях отличается от среднего, что может говорить о наличии выбросов.

 2.4 Оценка названий столбцов

In [7]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['Name', 'year', 'SellingPrice', 'kmdriven', 'fuel', 'seller_Type',
       'transmission', 'owner'],
      dtype='object')

In [ ]:
# переименование при необходимости

Названия столбцов понятные и без пробелов, поэтому в переименовании нет необходимости

### 3. Проверка пропусков

In [ ]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).



---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [8]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4339    False
4340     True
4341     True
4342    False
4343    False
Length: 4344, dtype: bool

Результат метода df.duplicated() показывает для каждой строки DataFrame, является ли она дубликатом предыдущей строки с такими же значениями (True), или уникальной (False).
В нашем случае для строк 4340 и 4341 ранее в списке элементов встречался дубликат.

In [16]:
# удалите дубликаты, если они есть
df = df.drop_duplicates().reset_index(drop=True) # удаляем явные дубликаты
df.duplicated() # проверяем результат

0       False
1       False
2       False
3       False
4       False
        ...  
3576    False
3577    False
3578    False
3579    False
3580    False
Length: 3581, dtype: bool

#### Проверка неявных дубликатов

In [26]:
df['year'].value_counts()

year
2017    347
2012    338
2015    333
2014    322
2013    298
2018    295
2016    283
2011    247
2010    211
2009    169
2019    162
2008    129
2007    117
2006     93
2005     61
2020     45
2004     38
2003     22
2002     18
2001     16
2000     12
1999      9
1998      9
1997      3
1996      2
1995      1
1992      1
Name: count, dtype: int64

In [29]:
df['fuel'].value_counts()

fuel
Diesel      1800
Petrol      1719
CNG           37
LPG           22
Petroll        1
Electric       1
Name: count, dtype: int64

Есть дубликат Petroll

In [30]:
df['fuel'] = df['fuel'].replace('Petroll', 'Petrol') # удаление дубликата
df['fuel'].value_counts() # проверка

fuel
Diesel      1800
Petrol      1720
CNG           37
LPG           22
Electric       1
Name: count, dtype: int64

Успешно удалили дубликат

In [31]:
df['seller_Type'].value_counts()

seller_Type
Individual          2836
Dealer               711
Trustmark Dealer      33
Dealeer                1
Name: count, dtype: int64

Есть дубликат Dealeer

In [32]:
df['seller_Type'] = df['seller_Type'].replace('Dealeer', 'Dealer') # удаление дубликата
df['seller_Type'].value_counts() # проверка

seller_Type
Individual          2836
Dealer               712
Trustmark Dealer      33
Name: count, dtype: int64

Успешно удалили дубликат

In [34]:
df['transmission'].value_counts()

transmission
Manual       3269
Automatic     312
Name: count, dtype: int64

Дубликатов нет

In [35]:
df['owner'].value_counts()

owner
First Owner             2220
Second Owner             980
Third Owner              289
Fourth & Above Owner      75
Test Drive Car            17
Name: count, dtype: int64

Дубликатов нет

In [ ]:
# удалите дубликаты, если они есть 

Все дубликаты удалены

---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [39]:
df.info() # проверяем типы данных присвоенные столбцам при чтении файла

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3581 entries, 0 to 3580
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          3581 non-null   object 
 1   year          3581 non-null   int64  
 2   SellingPrice  3579 non-null   float64
 3   kmdriven      3580 non-null   float64
 4   fuel          3580 non-null   object 
 5   seller_Type   3581 non-null   object 
 6   transmission  3581 non-null   object 
 7   owner         3581 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 223.9+ KB


In [42]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
df['kmdriven'] = df['kmdriven'].astype('Int64')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3581 entries, 0 to 3580
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          3581 non-null   object 
 1   year          3581 non-null   int64  
 2   SellingPrice  3579 non-null   float64
 3   kmdriven      3580 non-null   Int64  
 4   fuel          3580 non-null   object 
 5   seller_Type   3581 non-null   object 
 6   transmission  3581 non-null   object 
 7   owner         3581 non-null   object 
dtypes: Int64(1), float64(1), int64(1), object(5)
memory usage: 227.4+ KB


Был изменён тип данных столбца kmdriven для подсчёта пробега автомоболя в километрах с float на int.

In [43]:
df.isna().sum() # ищем пропуски

Name            0
year            0
SellingPrice    2
kmdriven        1
fuel            1
seller_Type     0
transmission    0
owner           0
dtype: int64

In [44]:
df = df.dropna(subset=['SellingPrice', 'kmdriven', 'fuel']) # удаляем все строки с пропусками
df.isna().sum() # проверяем удаление пропусков

Name            0
year            0
SellingPrice    0
kmdriven        0
fuel            0
seller_Type     0
transmission    0
owner           0
dtype: int64

Все пропуски успешно удалены

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Формулировка задания`*

Группировка - тип владельца (owner) по количеству разных коробок
передач (transmission)

In [56]:
# выполните группировку согласно варианту
df.groupby(['owner', 'transmission'])['transmission'].count()

owner                 transmission
First Owner           Automatic        227
                      Manual          1992
Fourth & Above Owner  Automatic          2
                      Manual            73
Second Owner          Automatic         64
                      Manual           915
Test Drive Car        Automatic          1
                      Manual            16
Third Owner           Automatic         18
                      Manual           271
Name: transmission, dtype: int64

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Основные наблюдения:
- Наибольшее число автомобилей — это автомобили у первого владельца с ручной коробкой передач (1992 шт.). Это говорит о популярности таких автомобилей среди тех, кто покупает машину "с нуля".
- Среди первого владельца также заметная доля "Automatic" (227), но их почти в 9 раз меньше, чем с ручной трансмиссией.
- У второго владельца аналогично: ручных машин (915) значительно больше, чем автоматических (64).
- В остальных категориях (Third Owner, Fourth & Above Owner, Test Drive Car) машины с ручной коробкой тоже преобладают над автоматом.
- Test Drive Car — довольно редкая категория, где всего 17 автомобилей (1 автомат и 16 ручных).
- В целом, машины с ручной трансмиссией встречаются гораздо чаще в каждом сегменте владельцев.

Вывод:
Рынок подержанных автомобилей данной выборки сильно смещён в сторону ручных трансмиссий, независимо от категории владельца. Автоматическая трансмиссия встречается заметно реже, особенно у автомобилей, переживших смену нескольких владельцев. Это может отражать предпочтения на рынке, доступность и стоимость обслуживания или специфику регионального спроса.



---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания`*

Группировка -тип коробки передач (transmission) по количеству разных
типов топлива (fuel). Создать датафрейм. Переименовать столбец с количеством в
“сount”.

In [63]:
# выполните группировку согласно варианту
gr2_series = df.groupby(['transmission', 'fuel'])['fuel'].count()
gr2_df = gr2_series.to_frame(name='count')
gr2_df

count
transmission fuel           
Automatic    Diesel      170
             Electric      1
             Petrol      141
Manual       CNG          37
             Diesel     1630
             LPG          22
             Petrol     1578

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Основные наблюдения:
- Среди автоматических коробок передач большинство автомобилей используют дизель (170) или бензин (141); электромобилей очень мало (1).
- Среди ручных коробок передач ярко выражено преобладание автомобилей на дизельном (1630) и бензиновом (1578) топливе.
- Альтернативные виды топлива (CNG, LPG) встречаются редко, только среди автомобилей с ручной коробкой (37 и 22 соответственно).

Вывод:
Основная масса автомобилей в выборке — это машины с ручной коробкой на дизельном или бензиновом топливе. Автомобили с автоматической трансмиссией встречаются гораздо реже и представлены в основном дизельными и бензиновыми моделями. Авто на альтернативном топливе (CNG, LPG, Electric) составляют незначительную часть рынка и почти все оснащены ручными коробками передач.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания`*

Сводная таблица (pivot_table) - средняя цена для каждого типа
владельца owner. Отсортировать по убыванию. Округлить до двух знаков

In [66]:
# выполните сводную таблицу согласно варианту

df.pivot_table(
    values='SellingPrice',
    index='owner',
    aggfunc='mean'
).sort_values(by='SellingPrice', ascending=False).round(2)

SellingPrice
owner                             
Test Drive Car           954293.94
First Owner              565513.71
Second Owner             341850.00
Third Owner              266142.21
Fourth & Above Owner     181213.29

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

Сводная таблица показывает, что максимальная средняя цена у автомобилей категории "Test Drive Car". Самые дорогие авто чаще всего принадлежат первому владельцу, а стоимость заметно снижается с каждым последующим владельцем. Минимальная средняя цена — у машин, находящихся в собственности четвертого и более владельца.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания`*

Сводная таблица (pivot_table) - среднее количество пройденных
километров для каждого года - строки и типа владельца - owner - столбцы.
Отсортировать по возрастанию year. Округлить до двух знаков

In [67]:
# выполните сводную таблицу согласно варианту

df.pivot_table(
    values='kmdriven',
    index='year',
    columns='owner',
    aggfunc='mean'
).sort_index().round(2)

owner  First Owner  Fourth & Above Owner  Second Owner  Test Drive Car  \
year                                                                     
1992          <NA>              100000.0          <NA>            <NA>   
1995          <NA>                  <NA>      100000.0            <NA>   
1996       60000.0                  <NA>       35000.0            <NA>   
1997       70000.0                  <NA>       80000.0            <NA>   
1998          <NA>               60000.0       76000.0            <NA>   
1999       70000.0               80000.0       65404.0            <NA>   
2000       77500.0               74250.0       77500.0            <NA>   
2001      106500.0               97000.0       95107.8            <NA>   
2002       76000.0               80000.0       87200.0            <NA>   
2003       55683.0              66666.67      98849.36            <NA>   
2004       79765.0              112750.0      91819.53            <NA>   
2005      79471.11               96322.5      79573.69            <NA>   
2006      95359.96              131848.8      94371.49            <NA>   
2007      86642.75             121555.56      87837.78            <NA>   
2008       79704.8               92500.0       86452.8            <NA>   
2009      97589.32              100000.0      94676.42            <NA>   
2010      78574.27              113000.0      98071.17            <NA>   
2011      81739.28              113175.0      96894.24            <NA>   
2012      82938.14             104333.33      86577.36            <NA>   
2013      73645.42               85000.0      81519.06            <NA>   
2014      68922.05              113000.0      85446.54            <NA>   
2015      61263.24              120000.0      69200.24            <NA>   
2016      56888.35                  <NA>      59843.57            <NA>   
2017      43108.69               50000.0      42005.85         24585.0   
2018      28434.67                  <NA>      30289.65         13000.0   
2019      19363.53                  <NA>       40400.0         7855.33   
2020      15084.38                1100.0          <NA>          790.33   

owner  Third Owner  
year                
1992          <NA>  
1995          <NA>  
1996          <NA>  
1997      120000.0  
1998       65000.0  
1999       50000.0  
2000       41371.5  
2001      64893.17  
2002       86000.0  
2003      100000.0  
2004      95971.38  
2005      119250.0  
2006     120940.91  
2007     101705.88  
2008     108434.48  
2009      98945.19  
2010     114209.68  
2011       99437.5  
2012       98896.4  
2013       82250.0  
2014     108166.67  
2015       82500.0  
2016       71600.0  
2017       57000.0  
2018       38900.0  
2019       15000.0  
2020          <NA>

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

Основные выводы:
- По мере уменьшения года выпуска авто (т.е. с развитием по строкам вверх), средний пробег возрастает у всех типов владельцев. Это ожидаемо, ведь старые автомобили эксплуатировались дольше и прошли больше километров.
- Для машин от первого владельца средний пробег постепенно снижается с 100 000 (старые года) до около 15 000 км для самых новых машин 2020 года. Это типичная ситуация: новые авто продаются с небольшим пробегом, а старые — с большим.
- У автомобилей «Fourth & Above Owner» пробег всегда выше, чем у машин первого или второго владельца — в среднем эта категория включает машины, которые уже активно эксплуатировались и часто переходили из рук в руки.
- Автомобили, проданные вторым владельцем, также имеют обычно значительный пробег, но чуть меньше, чем у четвертого и более владельца.
- «Test Drive Car» и «Third Owner» чаще имеют более низкий или нестабильный пробег, значения встречаются реже и зачастую ниже, что, возможно, связано с малым количеством таких машин в выборке.
- В целом чётко прослеживается тренд: чем старше год выпуска и чем больше хозяев было у машины — тем выше средний пробег.

Итог:
Таблица позволяет выявить зависимость пробега как от возраста авто, так и от типа владельца. Это важно для анализа рынка подержанных автомобилей — видно, что самые «свежие» и наименее эксплуатированные машины получают новые владельцы, а активно эксплуатируемые авто чаще переходят к следующим владельцам и имеют больший пробег.



---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

В ходе исследования анализировался набор данных, описывающий подержанные автомобили с такими характеристиками, как год выпуска, тип владельца, пробег, тип топлива, тип коробки передач, цена и другие параметры. Предметная область — рынок подержанных автомобилей.

Для анализа была проведена предварительная обработка данных, включающая:
- загрузку и чтение CSV-файла в pandas DataFrame,
- проверку наличия дублей и пропусков,
- изучение типов данных,
- преобразование типов данных для корректного анализа,
- систематический осмотр с помощью методов `info()` и `describe()`.

Дальнейший анализ включал группировки и построение сводных таблиц (`groupby()` и `pivot_table()`), где были исследованы следующие аспекты:

- Средняя цена автомобилей для разных категорий владельцев показала, что машины категории "Test Drive Car" и "First Owner" имеют самые высокие средние цены, а цена снижалась у машин с большим числом прежних владельцев.
- Распределение автомобилей по типам коробки передач и типам владельцев выявило, что большинство машин оснащены ручной коробкой передач, причем их значительно больше среди первых и вторых владельцев. Автомобили с автоматической трансмиссией и альтернативными видами топлива встречаются реже.
- Анализ пробега по годам выпуска и типам владельцев показал закономерность: чем старше автомобиль и чем больше у него владельцев, тем выше средний пробег автомобиля. Новые машины и тест-драйвовые автомобили имеют минимальный пробег соответственно.
- Были рассмотрены выбросы и вариативность значений средних, медианных показателей, что помогло понять, что цены и пробеги различаются значительно в зависимости от возраста и истории автомобиля.

Таким образом, исследование позволило структурировать и визуализировать основные характеристики рынка подержанных автомобилей, выявить подъемы и спады по ценам и пробегу в зависимости от категории владельцев, типов трансмиссии и топлива. Результаты могут быть полезны для оценки состояния автомобиля, планирования покупки-продажи и более глубокого понимания динамики вторичного автомобильного рынка.

### Дополнительное задание

**`Подробная формулировка задания`**

In [ ]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***